framework/src/main/java/org/checkerframework/framework/type/DefaultTypeHierarchy.java

```diff
@@ -285,104 +283,120 @@ protected boolean areEqualInHierarchy(
+   *
+   * <p>Containment is described in <a
    * href="https://docs.oracle.com/javase/specs/jls/se11/html/jls-4.html#jls-4.5.1">JLS section
    * 4.5.1 "Type Arguments of Parameterized Types"</a>.
    *
-   * @param inside a type argument of the "subtype"
-   * @param outside a type argument of the "supertype"
+   * <p>As described in <a
+   * href=https://docs.oracle.com/javase/specs/jls/se11/html/jls-4.html#jls-4.10.2>JLS section
+   * 4.10.2 Subtyping among Class and Interface Types</a>, a declared type is considered a supertype
+   * of another declared type only if all of the type arguments of the declared type "contain" the
+   * corresponding type arguments of the subtype.
+   *
+   * @param inside a possibly-contained type
+   * @param outside a possibly-containing type
    * @param canBeCovariant whether or not type arguments are allowed to be covariant
    * @return true if inside is contained by outside, or if canBeCovariant == true and {@code inside
    *     <: outside}
    */
   protected boolean isContainedBy(
-      final AnnotatedTypeMirror inside, final AnnotatedTypeMirror outside, boolean canBeCovariant) {
+      AnnotatedTypeMirror inside, AnnotatedTypeMirror outside, boolean canBeCovariant) {
 
-    if (ignoreUninferredTypeArgument(inside) || ignoreUninferredTypeArgument(outside)) {
+    if (shouldIgnoreUninferredTypeArgs(inside) || shouldIgnoreUninferredTypeArgs(outside)) {
       areEqualVisitHistory.put(inside, outside, currentTop, true);
       return true;
     }
 
-    if (outside.getKind() != TypeKind.WILDCARD
-        && !TypesUtils.isCaptured(outside.getUnderlyingType())) {
-      if (canBeCovariant) {
-        return isSubtype(inside, outside, currentTop);
+    if (outside.getKind() == TypeKind.WILDCARD) {
+      Boolean previousResult = areEqualVisitHistory.get(inside, outside, currentTop);
+      if (previousResult != null) {
+        return previousResult;
       }
-      return areEqualInHierarchy(inside, outside);
-    }
+      AnnotatedWildcardType outsideWildcard = (AnnotatedWildcardType) outside;
 
-    AnnotatedTypeMirror outsideUpperBound;
-    AnnotatedTypeMirror outsideLowerBound;
-    if (outside.getKind() == TypeKind.WILDCARD) {
-      outsideUpperBound = ((AnnotatedWildcardType) outside).getExtendsBound();
-      outsideLowerBound = ((AnnotatedWildcardType) outside).getSuperBound();
-    } else if (TypesUtils.isCaptured(outside.getUnderlyingType())) {
-      outsideUpperBound = ((AnnotatedTypeVariable) outside).getUpperBound();
-      outsideLowerBound = ((AnnotatedTypeVariable) outside).getLowerBound();
-    } else {
-      throw new BugInCF("Expected a wildcard or captured type variable, but found " + outside);
+      // Add a placeholder in case of recursion, to prevent infinite regress.
+      areEqualVisitHistory.put(inside, outside, currentTop, true);
+      boolean result =
+          isContainedByBoundType(
```

Comment:
Would a better name for this routine be `isContainedByBoundedWildcard`?

Reply:
I don't think so.  This method is called for unbound wildcards and also captured type variables.

Reply:
I wasn't sure what "BoundType" means.  It sounds like it's containment by the bound type itself, by it's really containment within the bounds.  Maybe `isContainedByBoundedType` or `isContainedWithinBounds` or the like?